### PR TITLE
add AllowanceCharge to HeaderTradeSettlement (solves #27)

### PR DESCRIPTION
**Pull request by blackgwe**

### CODE REVIEW

**Comment by PATROMO** [comment]
@blackgwe, thanks for the pull request. Can you please execute the command `make cs-fix` so that the code is formatted in a consistent way.

File: src/zugferd211/Model/AllowanceCharge.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+namespace Easybill\ZUGFeRD211\Model;
+
+use JMS\Serializer\Annotation as JMS;
+
+/**
+ * Class AllowanceCharge
+ *
+ * @package Easybill\ZUGFeRD\Model
+ */
+class AllowanceCharge
```
Comment (PATROMO) [inline src/zugferd211/Model/AllowanceCharge.php:13]: Please change the name to `TradeAllowanceCharge` as it is written in the XSD files. The suffix 'Type' is not used. 

image

**Review by PATROMO — CHANGES_REQUESTED** [review]
@blackgwe see my comments.

File: src/zugferd211/Model/AllowanceCharge.php (hunk 1 above)
Comment (blackgwe) [reply src/zugferd211/Model/AllowanceCharge.php:13]: I don't understand… Should I rename the class? In the `xsd` files e.g. [FACTUR-X_EN16931_…Entity_100.xsd](https://github.com/easybill/zugferd-php/blob/2890a319fc924dda43e14144429355c93297a181/src/zugferd211/Schema/EN16931/FACTUR-X_EN16931_urn_un_unece_uncefact_data_standard_ReusableAggregateBusinessInformationEntity_100.xsd) the complex types are always named with `Type` suffix, but your classes don't have this suffix , e.g. [Amount](https://github.com/easybill/zugferd-php/blob/master/src/zugferd211/Model/Amount.php) class. 
If you do not mind, please explain it to me.

Comment (PATROMO) [reply src/zugferd211/Model/AllowanceCharge.php:13]: @blackgwe I have added a suggestion. The suffiy `Type` we dont use.

**Comment by PATROMO** [comment]
@blackgwe can your please run `make cs-fix` or `./vendor/bin/php-cs-fixer fix ./src --config .php_cs.dist` to reformat your PHP files.

File: src/zugferd10/Builder.php [hunk 2]
```diff
@@ -25,7 +24,8 @@ public static function create(): Builder
     {
         $serializer = SerializerBuilder::create()
             ->setDebug(true)
-            ->build();
+            ->build()
```
Comment (blackgwe) [inline src/zugferd10/Builder.php:27]: I've not used php-cs-fixer before, I'm not satisfied with this result…
`./vendor/bin/php-cs-fixer fix ./src --config .php_cs.dist`

**Comment by PATROMO** [comment]
@blackgwe sorry for the mess. The `master` was not tidy. I have now cleaned up the master (5fc7fa7). Can you pleas pull and merge the current master into your pull request. Here you can read something about the tool php-cs-fix https://github.com/FriendsOfPHP/PHP-CS-Fixer .

**Comment by blackgwe** [comment]
to clean up the mess, I provide my changes in a single commit.

File: src/zugferd211/Model/TradeAllowanceCharge.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * Class AllowanceCharge.
+ */
+class TradeAllowanceCharge
+{
+    /**
+     * @JMS\Type("Easybill\ZUGFeRD211\Model\Indicator")
+     * @JMS\XmlElement(cdata=false, namespace="urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100")
+     * @JMS\SerializedName("ChargeIndicator")
+     */
+    public Indicator $indicator;
```
Comment (PATROMO) [inline src/zugferd211/Model/TradeAllowanceCharge.php:19]:
```suggestion
    public ?Indicator $indicator = null;
```

File: src/zugferd211/Model/TradeAllowanceCharge.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+     * @JMS\Type("Easybill\ZUGFeRD211\Model\Amount")
+     * @JMS\XmlElement(cdata=false, namespace="urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100")
+     * @JMS\SerializedName("BasisAmount")
+     */
+    public Amount $basisAmount;
```
Comment (PATROMO) [inline src/zugferd211/Model/TradeAllowanceCharge.php:26]:
```suggestion
    public ?Amount $basisAmount = null;
```

File: src/zugferd211/Model/TradeAllowanceCharge.php [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+     * @JMS\SerializedName("ActualAmount")
+     */
+    public Amount $actualAmount;
+
+    /**
+     * @JMS\Type("string")
+     * @JMS\XmlElement(cdata = false, namespace = "urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100")
+     * @JMS\SerializedName("Reason")
+     */
+    public string $reason;
```
Comment (PATROMO) [inline src/zugferd211/Model/TradeAllowanceCharge.php:40]:
```suggestion
    public ?string $reason = null;
```

File: src/zugferd211/Model/TradeAllowanceCharge.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    /**
+     * @var TradeTax[]
+     * @JMS\Type("array<Easybill\ZUGFeRD211\Model\TradeTax>")
+     * @JMS\XmlList(inline = true, entry = "CategoryTradeTax", namespace="urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100")
+     */
+    public array $tradeTax;
```
Comment (PATROMO) [inline src/zugferd211/Model/TradeAllowanceCharge.php:47]:
```suggestion
    public array $tradeTax = [];
```

**Review by PATROMO — CHANGES_REQUESTED** [review]
@blackgwe it looks good now. One more thing: In TradeAllowanceCharge not all fields are required. Have a look at my code suggestions.

File: src/zugferd211/Model/TradeAllowanceCharge.php (hunk 6 above)
Comment (blackgwe) [reply src/zugferd211/Model/TradeAllowanceCharge.php:47]: The $actualAmount is not required either, is it?

Comment (PATROMO) [reply src/zugferd211/Model/TradeAllowanceCharge.php]: $actualAmount is required.
Bildschirmfoto 2020-11-16 um 09 59 38

**Review by PATROMO — APPROVED** [review]

**Pull request merged**